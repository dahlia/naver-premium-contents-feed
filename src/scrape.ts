export interface ChannelId {
  cpName: string;
  subId: string;
}

export interface Channel {
  url: URL;
  name: string;
  description: string;
  partner: boolean;
  provider: null | {
    name: string;
    email: string | null;
  };
  thumbnailUrl: URL;
  coverImageUrl: URL;
  categories: Record<CategoryId, Category>;
  latestUpdated: Date;
}

export type CategoryId = string;

export interface Category {
  id: CategoryId;
  name: string;
  url: URL;
}

function getUrl(channelId: ChannelId): URL | string {
  return "https://contents.premium.naver.com/ch/template/SCS_PREMIUM_CONTENT_LIST" +
    `?cpName=${channelId.cpName}&subId=${channelId.subId}`;
}

export async function scrape(
  channelId: ChannelId,
  urlGetter: (channelId: ChannelId) => URL | string = getUrl,
): Promise<Channel> {
  const url = urlGetter(channelId);
  const response = await fetch(url);
  const data = await response.json();
  const channelInfo = data.component.SCS_PREMIUM_CHANNEL_INFO_V1.value;
  const providerName: string | undefined = channelInfo.provider ??
    channelInfo.representativeName?.trim();
  const cpRegisterInfo = channelInfo.cpInfo?.cpRegisterInfo;
  const categoryList: {
    categoryId: string;
    categoryName: string;
    contentListByCategoryIdUrl: string;
  }[] = data.component.SCS_PREMIUM_CATEGORY_LIST_V1.value.data;
  const categories: Record<CategoryId, Category> = Object.fromEntries(
    categoryList
      .filter((c) => c.categoryId !== "")
      .map((c) => [c.categoryId, {
        id: c.categoryId,
        name: c.categoryName,
        url: new URL(
          c.contentListByCategoryIdUrl,
          channelInfo.absoluteHomeUrl,
        ),
      }]),
  );
  return {
    url: new URL(channelInfo.absoluteHomeUrl),
    name: channelInfo.channelName,
    description: channelInfo.description,
    partner: channelInfo.isPartner,
    provider: providerName != null && providerName.trim() != ""
      ? {
        name: channelInfo.provider ??
          (channelInfo.representativeName?.trim() || undefined),
        email: cpRegisterInfo?.cpTitle?.trim() === providerName
          ? cpRegisterInfo?.email?.trim() ?? null
          : null,
      }
      : null,
    thumbnailUrl: new URL(channelInfo.thumbnail),
    coverImageUrl: new URL(channelInfo.coverImage),
    categories,
    latestUpdated: new Date(
      channelInfo.channelInfo.lastContentPublishDt + "+09:00",
    ),
  };
}
