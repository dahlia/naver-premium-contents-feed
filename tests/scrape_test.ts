import { assertSnapshot } from "std/testing/snapshot.ts";
import { ChannelId, scrape } from "../src/scrape.ts";

function getUrl(channelId: ChannelId): URL | string {
  return new URL(
    `data/contents-${channelId.cpName}-${channelId.subId}.json`,
    import.meta.url,
  );
}

Deno.test("scrape()", async (ctx) => {
  const nonPartner = await scrape(
    { cpName: "historia9110", subId: "historia91" },
    getUrl,
  );
  await assertSnapshot(ctx, nonPartner);
  const partner = await scrape(
    { cpName: "astrotales", subId: "knowledge" },
    getUrl,
  );
  await assertSnapshot(ctx, partner);
});
