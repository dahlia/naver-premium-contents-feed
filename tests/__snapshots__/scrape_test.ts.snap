export const snapshot = {};

snapshot[`scrape() 1`] = `
{
  categories: {
    "17f256afade00085s": {
      id: "17f256afade00085s",
      name: "혁명 읽는 사람",
      url: URL {
        hash: "",
        host: "contents.premium.naver.com",
        hostname: "contents.premium.naver.com",
        href: "https://contents.premium.naver.com/historia9110/historia91/contents?categoryId=17f256afade00085s",
        origin: "https://contents.premium.naver.com",
        password: "",
        pathname: "/historia9110/historia91/contents",
        port: "",
        protocol: "https:",
        search: "?categoryId=17f256afade00085s",
        username: "",
      },
    },
    "17feaa9cec5000qlg": {
      id: "17feaa9cec5000qlg",
      name: "서평글 모음",
      url: URL {
        hash: "",
        host: "contents.premium.naver.com",
        hostname: "contents.premium.naver.com",
        href: "https://contents.premium.naver.com/historia9110/historia91/contents?categoryId=17feaa9cec5000qlg",
        origin: "https://contents.premium.naver.com",
        password: "",
        pathname: "/historia9110/historia91/contents",
        port: "",
        protocol: "https:",
        search: "?categoryId=17feaa9cec5000qlg",
        username: "",
      },
    },
    "1800259cf41000g4h": {
      id: "1800259cf41000g4h",
      name: "시평",
      url: URL {
        hash: "",
        host: "contents.premium.naver.com",
        hostname: "contents.premium.naver.com",
        href: "https://contents.premium.naver.com/historia9110/historia91/contents?categoryId=1800259cf41000g4h",
        origin: "https://contents.premium.naver.com",
        password: "",
        pathname: "/historia9110/historia91/contents",
        port: "",
        protocol: "https:",
        search: "?categoryId=1800259cf41000g4h",
        username: "",
      },
    },
    "1825ce06e90000ugd": {
      id: "1825ce06e90000ugd",
      name: "영화후기",
      url: URL {
        hash: "",
        host: "contents.premium.naver.com",
        hostname: "contents.premium.naver.com",
        href: "https://contents.premium.naver.com/historia9110/historia91/contents?categoryId=1825ce06e90000ugd",
        origin: "https://contents.premium.naver.com",
        password: "",
        pathname: "/historia9110/historia91/contents",
        port: "",
        protocol: "https:",
        search: "?categoryId=1825ce06e90000ugd",
        username: "",
      },
    },
  },
  coverImageUrl: URL {
    hash: "",
    host: "scs-phinf.pstatic.net",
    hostname: "scs-phinf.pstatic.net",
    href: "https://scs-phinf.pstatic.net/MjAyMjAyMjNfMjk4/MDAxNjQ1NTg3NDI3OTA2.allrMN9KTJsBnhWZTngyzyS2MnNDngpifptY0isPwY4g.TXyGHg-Vs3Bxh0VF1a7v8JdAmKEGCxR0QY75Qfx4VH0g.JPEG/image%7Cpremium%7Cchannel%7Chistoria9110%7C2022%7C02%7C23%7C1645587427877.jpg",
    origin: "https://scs-phinf.pstatic.net",
    password: "",
    pathname: "/MjAyMjAyMjNfMjk4/MDAxNjQ1NTg3NDI3OTA2.allrMN9KTJsBnhWZTngyzyS2MnNDngpifptY0isPwY4g.TXyGHg-Vs3Bxh0VF1a7v8JdAmKEGCxR0QY75Qfx4VH0g.JPEG/image%7Cpremium%7Cchannel%7Chistoria9110%7C2022%7C02%7C23%7C1645587427877.jpg",
    port: "",
    protocol: "https:",
    search: "",
    username: "",
  },
  description: "오늘의 시대에서 혁명을 읽어낸다. 20세기 레닌의 시각과 21세기 우리의 시각을 교차시키는 책읽기. 책서평, 정치비평 및 레닌에 관한 연구 연재.",
  latestUpdated: 2023-01-01T15:00:34.903Z,
  name: "혁명 읽는 사람",
  partner: false,
  provider: {
    email: "fpdlakstp@naver.com",
    name: "손민석",
  },
  thumbnailUrl: URL {
    hash: "",
    host: "scs-phinf.pstatic.net",
    hostname: "scs-phinf.pstatic.net",
    href: "https://scs-phinf.pstatic.net/MjAyMjAyMjNfNjcg/MDAxNjQ1NTg3MTcyODM1.nIbCznb-zkIbxNyNomKqbEWws-bboqE4irvcD958b3wg.TFgVzhUbFQAUf_MSjTKjLAdx188eRlLObbKJmMtHhWEg.PNG/image%7Cpremium%7Cchannel%7Chistoria9110%7C2022%7C02%7C23%7C1645587172809.png?type=nfs200_200",
    origin: "https://scs-phinf.pstatic.net",
    password: "",
    pathname: "/MjAyMjAyMjNfNjcg/MDAxNjQ1NTg3MTcyODM1.nIbCznb-zkIbxNyNomKqbEWws-bboqE4irvcD958b3wg.TFgVzhUbFQAUf_MSjTKjLAdx188eRlLObbKJmMtHhWEg.PNG/image%7Cpremium%7Cchannel%7Chistoria9110%7C2022%7C02%7C23%7C1645587172809.png",
    port: "",
    protocol: "https:",
    search: "?type=nfs200_200",
    username: "",
  },
  url: URL {
    hash: "",
    host: "contents.premium.naver.com",
    hostname: "contents.premium.naver.com",
    href: "https://contents.premium.naver.com/historia9110/historia91",
    origin: "https://contents.premium.naver.com",
    password: "",
    pathname: "/historia9110/historia91",
    port: "",
    protocol: "https:",
    search: "",
    username: "",
  },
}
`;

snapshot[`scrape() 2`] = `
{
  categories: {
    "17e385420c8000erx": {
      id: "17e385420c8000erx",
      name: "망원경 이야기",
      url: URL {
        hash: "",
        host: "contents.premium.naver.com",
        hostname: "contents.premium.naver.com",
        href: "https://contents.premium.naver.com/astrotales/knowledge/contents?categoryId=17e385420c8000erx",
        origin: "https://contents.premium.naver.com",
        password: "",
        pathname: "/astrotales/knowledge/contents",
        port: "",
        protocol: "https:",
        search: "?categoryId=17e385420c8000erx",
        username: "",
      },
    },
    "17e38546e5e000aht": {
      id: "17e38546e5e000aht",
      name: "천문학자 이야기",
      url: URL {
        hash: "",
        host: "contents.premium.naver.com",
        hostname: "contents.premium.naver.com",
        href: "https://contents.premium.naver.com/astrotales/knowledge/contents?categoryId=17e38546e5e000aht",
        origin: "https://contents.premium.naver.com",
        password: "",
        pathname: "/astrotales/knowledge/contents",
        port: "",
        protocol: "https:",
        search: "?categoryId=17e38546e5e000aht",
        username: "",
      },
    },
    "17e5bff79ec000aia": {
      id: "17e5bff79ec000aia",
      name: "관측 이야기",
      url: URL {
        hash: "",
        host: "contents.premium.naver.com",
        hostname: "contents.premium.naver.com",
        href: "https://contents.premium.naver.com/astrotales/knowledge/contents?categoryId=17e5bff79ec000aia",
        origin: "https://contents.premium.naver.com",
        password: "",
        pathname: "/astrotales/knowledge/contents",
        port: "",
        protocol: "https:",
        search: "?categoryId=17e5bff79ec000aia",
        username: "",
      },
    },
    "18118ba9dc9000wjj": {
      id: "18118ba9dc9000wjj",
      name: "천문학 이야기",
      url: URL {
        hash: "",
        host: "contents.premium.naver.com",
        hostname: "contents.premium.naver.com",
        href: "https://contents.premium.naver.com/astrotales/knowledge/contents?categoryId=18118ba9dc9000wjj",
        origin: "https://contents.premium.naver.com",
        password: "",
        pathname: "/astrotales/knowledge/contents",
        port: "",
        protocol: "https:",
        search: "?categoryId=18118ba9dc9000wjj",
        username: "",
      },
    },
    "1822f1eeba6000wva": {
      id: "1822f1eeba6000wva",
      name: "천문늬우스",
      url: URL {
        hash: "",
        host: "contents.premium.naver.com",
        hostname: "contents.premium.naver.com",
        href: "https://contents.premium.naver.com/astrotales/knowledge/contents?categoryId=1822f1eeba6000wva",
        origin: "https://contents.premium.naver.com",
        password: "",
        pathname: "/astrotales/knowledge/contents",
        port: "",
        protocol: "https:",
        search: "?categoryId=1822f1eeba6000wva",
        username: "",
      },
    },
    "182536d8359000br2": {
      id: "182536d8359000br2",
      name: "JWST 이야기",
      url: URL {
        hash: "",
        host: "contents.premium.naver.com",
        hostname: "contents.premium.naver.com",
        href: "https://contents.premium.naver.com/astrotales/knowledge/contents?categoryId=182536d8359000br2",
        origin: "https://contents.premium.naver.com",
        password: "",
        pathname: "/astrotales/knowledge/contents",
        port: "",
        protocol: "https:",
        search: "?categoryId=182536d8359000br2",
        username: "",
      },
    },
  },
  coverImageUrl: URL {
    hash: "",
    host: "scs-phinf.pstatic.net",
    hostname: "scs-phinf.pstatic.net",
    href: "https://scs-phinf.pstatic.net/MjAyMTEyMDhfNjMg/MDAxNjM4OTQ4NzM5Njk4.3EU-hzwg--ipbzTldW-xNw8uAG-wmueiCMZNwmS1F-Ag.SgYkUTDL96EM0isirgjj6g8DFyaoEnr3UIltW8-YEKMg.PNG/image%7Cpremium%7Cchannel%7Castrotales%7C2021%7C12%7C08%7C1638948739661.png",
    origin: "https://scs-phinf.pstatic.net",
    password: "",
    pathname: "/MjAyMTEyMDhfNjMg/MDAxNjM4OTQ4NzM5Njk4.3EU-hzwg--ipbzTldW-xNw8uAG-wmueiCMZNwmS1F-Ag.SgYkUTDL96EM0isirgjj6g8DFyaoEnr3UIltW8-YEKMg.PNG/image%7Cpremium%7Cchannel%7Castrotales%7C2021%7C12%7C08%7C1638948739661.png",
    port: "",
    protocol: "https:",
    search: "",
    username: "",
  },
  description: "천문학자가 들려주는 그 누구도 알려주지 않던
신기하고 은밀한 천문학 이야기, 천일야화",
  latestUpdated: 2022-11-30T14:15:22.628Z,
  name: "신기하고 은밀한 천문학이야기 천일야화",
  partner: true,
  provider: null,
  thumbnailUrl: URL {
    hash: "",
    host: "scs-phinf.pstatic.net",
    hostname: "scs-phinf.pstatic.net",
    href: "https://scs-phinf.pstatic.net/MjAyMTEyMDhfMTU3/MDAxNjM4OTQ4NzM1NjE4.oLyKgLw9NOIo0NIng4qh21mxRj1neHC_UPRKvZRmSx4g.Om9fnQWd7TMKeplXmOH2XHwCgKpHLdm8K0eoe8I2al0g.PNG/image%7Cpremium%7Cchannel%7Castrotales%7C2021%7C12%7C08%7C1638948735231.png?type=nfs200_200",
    origin: "https://scs-phinf.pstatic.net",
    password: "",
    pathname: "/MjAyMTEyMDhfMTU3/MDAxNjM4OTQ4NzM1NjE4.oLyKgLw9NOIo0NIng4qh21mxRj1neHC_UPRKvZRmSx4g.Om9fnQWd7TMKeplXmOH2XHwCgKpHLdm8K0eoe8I2al0g.PNG/image%7Cpremium%7Cchannel%7Castrotales%7C2021%7C12%7C08%7C1638948735231.png",
    port: "",
    protocol: "https:",
    search: "?type=nfs200_200",
    username: "",
  },
  url: URL {
    hash: "",
    host: "contents.premium.naver.com",
    hostname: "contents.premium.naver.com",
    href: "https://contents.premium.naver.com/astrotales/knowledge",
    origin: "https://contents.premium.naver.com",
    password: "",
    pathname: "/astrotales/knowledge",
    port: "",
    protocol: "https:",
    search: "",
    username: "",
  },
}
`;
